File: impl/src/org/ufomatics/archipelago/view/swing/DatasourcePane.java
package org.ufomatics.archipelago.view.swing;

import org.ufomatics.archipelago.control.UFOPlatformController;
import org.ufomatics.archipelago.model.DatabaseAdapter;
import org.ufomatics.archipelago.model.Database;

import javax.swing.*;
import java.util.List;
import java.util.ArrayList;

public class DatasourcePane extends JPanel {
    private JTextField nameTextfield;
    private JComboBox adaptersCombo;
    private UFOPlatformController controller;
    private Database datasource;

    public DatasourcePane(UFOPlatformController controller, Database datasource) {
        SpringLayout layout = new SpringLayout();
        setLayout(layout);

        this.controller = controller;
        this.datasource = datasource;

        createAdaptersCombo(controller);

        JLabel nameLabel = new JLabel("Name");
        nameTextfield = new JTextField(datasource.getName());
        JLabel typeLabel = new JLabel("Adapter");
        adaptersCombo = createAdaptersCombo(controller);
        DatabaseAdapter selectedType = datasource.getAdapter() == null ? controller.getDefaultAdapter() : datasource.getAdapter();
        adaptersCombo.setSelectedItem(selectedType);
        add(nameLabel);
        add(nameTextfield);
        add(typeLabel);
        add(adaptersCombo);
        layout.putConstraint(SpringLayout.EAST, this, 5, SpringLayout.EAST, adaptersCombo);
        layout.putConstraint(SpringLayout.SOUTH, this, 5, SpringLayout.SOUTH, adaptersCombo);
        layout.putConstraint(SpringLayout.EAST, nameTextfield, 5, SpringLayout.EAST, adaptersCombo);
        layout.putConstraint(SpringLayout.WEST, nameLabel, 5, SpringLayout.WEST, this);
        layout.putConstraint(SpringLayout.NORTH, nameLabel, 5, SpringLayout.NORTH, this);
        layout.putConstraint(SpringLayout.WEST, nameTextfield, 5, SpringLayout.EAST, typeLabel);
        layout.putConstraint(SpringLayout.NORTH, nameTextfield, 0, SpringLayout.NORTH, nameLabel);
        layout.putConstraint(SpringLayout.WEST, typeLabel, 0, SpringLayout.WEST, nameLabel);
        layout.putConstraint(SpringLayout.NORTH, typeLabel, 5, SpringLayout.SOUTH, nameLabel);
        layout.putConstraint(SpringLayout.WEST, adaptersCombo, 5, SpringLayout.EAST, typeLabel);
        layout.putConstraint(SpringLayout.NORTH, adaptersCombo, 5, SpringLayout.SOUTH, nameTextfield);
    }

    private JComboBox createAdaptersCombo(UFOPlatformController controller) {
        JComboBox adaptersCombo = new JComboBox();
        final List<String> adapters = new ArrayList<String>();
        for (DatabaseAdapter adapter : controller.getAdapters()) {
            adapters.add(adapter.getName());
        }
        adaptersCombo.setModel(new DefaultComboBoxModel(){
            public Object getElementAt(int index) {
                return adapters.get(index);
            }
            public int getSize() {
                return adapters.size();
            }
        });
        return adaptersCombo;
    }

    public void ok() {
        Object selectedItem = adaptersCombo.getSelectedItem();
        for (DatabaseAdapter adapter : controller.getAdapters()) {
            if (selectedItem.equals (adapter.getName())) {
                datasource.setAdapter(adapter);
                break;
            }
        }
        datasource.setName(nameTextfield.getText());
    }
}

File: src/archipelago/impl/src/ufomatics/archipelago/view/swing/DatasourcePane.ts
import {JPanel} from "archipelago/api/util/jsdk/swing/JPanel"
import {JTextField} from "archipelago/api/util/jsdk/swing/JTextField"
import {Database} from "archipelago/api/model/Database"
import {JComboBox} from "archipelago/api/util/jsdk/swing/JComboBox"
import {UFOPlatformController} from "archipelago/api/control/UFOPlatformController"
import {JLabel} from "archipelago/api/util/jsdk/swing/JLabel"
import {DefaultComboBoxModel} from "archipelago/api/util/jsdk/swing/DefaultComboBoxModel"
import {SpringLayout, SpringLayoutDirection} from "archipelago/api/util/jsdk/swing/SpringLayout"

export class DatasourcePane extends JPanel {
  private nameTextfield: JTextField
  private adaptersCombo: JComboBox

  constructor(protected controller: UFOPlatformController, protected datasource: Database) {
    super()
    const layout = new SpringLayout()
    this.setLayout(layout)

    const nameLabel = new JLabel("Name")
    this.nameTextfield = new JTextField(datasource.getName())
    const typeLabel = new JLabel("Adapter")
    this.adaptersCombo = this.createAdaptersCombo(controller)
    const selectedType = datasource.getAdapter() == null ? controller.getDefaultAdapter() : datasource.getAdapter()
    this.adaptersCombo.setSelectedItem(selectedType)
    this.add(nameLabel)
    this.add(this.nameTextfield)
    this.add(typeLabel)
    this.add(this.adaptersCombo)
    layout.putConstraint(SpringLayoutDirection.EAST, this, 5, SpringLayoutDirection.EAST, this.adaptersCombo)
    layout.putConstraint(SpringLayoutDirection.SOUTH, this, 5, SpringLayoutDirection.SOUTH, this.adaptersCombo)
    layout.putConstraint(SpringLayoutDirection.EAST, this.nameTextfield, 5, SpringLayoutDirection.EAST, this.adaptersCombo)
    layout.putConstraint(SpringLayoutDirection.WEST, nameLabel, 5, SpringLayoutDirection.WEST, this)
    layout.putConstraint(SpringLayoutDirection.NORTH, nameLabel, 5, SpringLayoutDirection.NORTH, this)
    layout.putConstraint(SpringLayoutDirection.WEST, this.nameTextfield, 5, SpringLayoutDirection.EAST, typeLabel)
    layout.putConstraint(SpringLayoutDirection.NORTH, this.nameTextfield, 0, SpringLayoutDirection.NORTH, nameLabel)
    layout.putConstraint(SpringLayoutDirection.WEST, typeLabel, 0, SpringLayoutDirection.WEST, nameLabel)
    layout.putConstraint(SpringLayoutDirection.NORTH, typeLabel, 5, SpringLayoutDirection.SOUTH, nameLabel)
    layout.putConstraint(SpringLayoutDirection.WEST, this.adaptersCombo, 5, SpringLayoutDirection.EAST, typeLabel)
    layout.putConstraint(SpringLayoutDirection.NORTH, this.adaptersCombo, 5, SpringLayoutDirection.SOUTH, this.nameTextfield)
  }

  ok(): void {
    const selectedItem = this.adaptersCombo.getSelectedItem()
    for (const adapter of this.controller.getAdapters()._set) {
      if (selectedItem.equals(adapter.getName())) {
        this.datasource.setAdapter(adapter)
        break
      }
    }
    this.datasource.setName(this.nameTextfield.getText())
  }

  private createAdaptersCombo(controller: UFOPlatformController): JComboBox {
    const adaptersCombo = new JComboBox()
    const adapters: string[] = []
    for (const adapter of controller.getAdapters()._set) {
      adapters.push(adapter.getName())
    }
    adaptersCombo.setModel(new class extends DefaultComboBoxModel {
      getElementAt(index: number): any {
        return adapters[index]
      }

      getSize(): number {
        return adapters.length
      }
    }())
    return adaptersCombo
  }
}
